### ISSUE TITLE
Best practice for hosting niivue app on GitHub pages with Vite

### ISSUE THREAD

**Issue by kolibril13** [niivue/niivue#827]
Hi there!
today I had the chance to make some experiments with niivue, and here I want to share some things I've come up with.
First of all I've tried the [react app example](https://github.com/niivue/niivue/blob/main/DEVELOP.md#react-example) and that works out of the box:
Repo: https://github.com/kolibril13/niivue-vite-react-app
Page: https://kolibril13.github.io/niivue-vite-react-app/

next I've tried the [basic vanilla JS example](https://github.com/niivue/niivue/blob/main/DEVELOP.md#load-a-volume) and that worked great as well.   
However, here I had to first create a canvas element to make this work.

Repo: https://github.com/kolibril13/niivue-vite-vanillajs-app
Page: https://kolibril13.github.io/niivue-vite-vanillajs-app/

My question regarding the Vanilla JS example: 
Is this how Niivue should be used, or can the below example still be improved?
Looking forward to feedback!
```js
import { Niivue } from "@niivue/niivue";

// Create a canvas element
const canvas = document.createElement("canvas");
canvas.id = "gl";
canvas.width = 800;
canvas.height = 800;

// Append the canvas to the element with id="app"
document.querySelector("#app").appendChild(canvas);

// Make an array of volumes to load
let volumeList = [
    { url: "https://niivue.github.io/niivue-demo-images/mni152.nii.gz" },
];

const nv = new Niivue({ isResizeCanvas: false });
nv.attachTo("gl"); // attach to canvas with id="gl"
nv.loadVolumes(volumeList);
```

**Comment by hanayik** [comment]
The vanilla JS example looks good to me. The only alternative would be that the canvas already exists in the html page and the JavaScript only attaches NiiVue to the canvas rather than creating the canvas in JS.

**Comment by kolibril13** [comment]
That's great to hear!
For **ipyNiiVue** was translating the above example into an anywidget widget by replacing `document.querySelector("#app").appendChild(canvas)` with `  el.appendChild(canvas);` like this:
```js
import { Niivue } from "@niivue/niivue";

export function render({ model, el }) {
  const canvas = document.createElement("canvas");
  canvas.id = "gl";
  canvas.width = 800;
  canvas.height = 800;

  // Append the canvas to the element with id="app"
  el.appendChild(canvas);

  // Make an array of volumes to load
  let volumeList = [
    { url: "https://niivue.github.io/niivue-demo-images/mni152.nii.gz" },
  ];

  //
  const nv = new Niivue({ isResizeCanvas: false });
  nv.attachTo("gl"); // attach to canvas with id="gl"
  nv.loadVolumes(volumeList);
}
```
Unfortunately, I now get an error:
```console
ncaught (in promise) TypeError: Cannot read properties of null (reading 'getContext')
```

Interestingly, the error goes away, and the widget **does appear** when I trigger the Hot Module Replacement. 
Triggering Hot Module Replacement (HMR) means saving the changes and letting npm rebuild the JS without restarting the notebook and without reloading the page, here's a screen recording of this:

https://github.com/niivue/niivue/assets/44469195/9d1b90af-d900-4a19-9810-0200c96f38c9

Probably that has to do something with the way the canvas is added to the widget output and the async call of `loadVolumes` (as also mentioned here https://github.com/niivue/ipyniivue/issues/44#issuecomment-1880972855)

Further notes: 
* This error does only occurs in the anywidget vanillaJS version: 
https://github.com/kolibril13/anywidget-ipyniivue-vanillajs/blob/main/js/widget.js
* This error **does not** occur at all in the anywidget react version:  
https://github.com/kolibril13/anywidget-ipyniivue/blob/main/js/widget.jsx

**Any ideas?** 

cc. @christian-oreilly , @manzt , @cdrake

**Comment by neurolabusc** [comment]
@kolibril13 javascript runs in parallel by design, but one must wait until your canvas exists before you can get the WebGL context from the context. So your problem is you can not attach a NiiVue instance to a canvas until the canvas is created, so you must await the completion of `document.createElement("canvas")` before you call `nv.attachTo("gl")`.

Several solutions are [described here](https://stackoverflow.com/questions/10291693/cannot-read-property-getcontext-of-null-using-canvas) but I am not sure what the most elegant method for anywidget work flow. 

As an aside, you will want to ensure that `isResizeCanvas: false` is the correct for anywidget - this enforces the idea that the canvas has a fixed resolution that will not change to fit a container.

One final note, some computations are very slow to compute and so NiiVue will defer running them until requested. This is unintuitive, but it does allow you to change a large number of properties before waiting for a slow computation. This is described in [issue 463](https://github.com/niivue/niivue/issues/463). The call `nv1.updateGLVolume();` forces NiiVue to apply all the deferred changes.

**Comment by manzt** [comment]
@kolibril13, why not use a different api that accepts the canvas element directly (`attachToCanvas`)?:

```js
import { Niivue } from "@niivue/niivue";

export async function render({ model, el }) {
	let canvas = document.createElement("canvas");
	let container = document.createElement("div");
	container.style.height = "300px";
	container.appendChild(canvas);
	el.appendChild(container);

	let volumeList = [{ url: "https://niivue.github.io/niivue-demo-images/mni152.nii.gz" }];
	let nv = new Niivue();
	nv.attachToCanvas(canvas);
	nv.loadVolumes(volumeList);
	return () => {};
}
```

**Comment by kolibril13** [comment]
`nv.attachToCanvas(canvas);` solved this problem 🎉 

Also thanks a lot for the video!
I especially like this part at [in the middle](https://youtu.be/8aGhZQkoFbQ?si=DODtOwpuLSo1HJHn&t=768) where the terms `stack` `webapis`, `task queue` and `event loop` are explained in an intuitive visual way.
<img width="1016" alt="image" src="https://github.com/niivue/niivue/assets/44469195/36a907d1-d415-4c43-a9c2-47d946cd77e1">
